### ISSUE TITLE
Detected application deadlock on message

### ISSUE THREAD

**Issue by arbesoft** [dotnet/orleans#6646]
prerequisite: PerformDeadlockDetection = true;

I need some help, I use Azure Event Hub for Streams. Following situation, GrainA publishes a message and GrainB receive it, do any jobs and publish an update about his state. GrainA can not receive this update because get a Deadlock exception. 

Isn't that possible?

Thx for your help

**Comment by sergeybykov** [comment]
If you use an external queue, such as EventHub, for streaming, the grain that publishes an event succeeds with that when a stream event is written to the queue, but before it is actually delivered and processed by subscribers to that stream. Because of that, you shouldn't get any deadlocks in the scenario you described.

Such deadlocks are possible with the SMS stream provider that doesn't use any external queues and instead delivers events directly as grain calls between grains.

**Comment by arbesoft** [comment]
Thx for your answer. That was also what I was thinking about streams with EventHub. But then I got this logs. I will search perhaps there is a other problem. :(


{
  "_index": "2020.07.13",
  "_type": "_doc",
  "_id": "94yFRXMBxwDa4XbtRq1o",
  "_score": 1,
  "_source": {
    "@timestamp": "2020-07-13T00:13:47.3171111+00:00",
    "level": "Warning",
    "messageTemplate": "Detected application deadlock on message {Message} and activation {Activation}",
    "message": "Detected application deadlock on message \"Request S10.244.0.181:11111:332295207*stg/7872883655142754957/d6c34a8d@Sd6c34a8d->S10.244.0.181:11111:332295207*grn/E4F58DB0/9ccbb316@6285d9a0 #260\" and activation \"[Activation: S10.244.0.181:11111:332295207*grn/E4F58DB0/9ccbb316@6285d9a0 #GrainType=Grains.MemberGrain Placement=RandomPlacement State=Valid]\"",
    "exceptions": [
      {
        "Depth": 0,
        "ClassName": "Orleans.Runtime.DeadlockException",
        "Message": "Deadlock Exception for grain call chain [*grn/E4F58DB0/9ccbb316, *grn/8206F545/7e16f2b1, *grn/E4F58DB0/9ccbb316].",
        "Source": "Orleans.Runtime",
        "StackTraceString": "   at Orleans.Runtime.Dispatcher.CheckDeadlock(Message message)\n   at Orleans.Runtime.Dispatcher.ReceiveRequest(Message message, ActivationData targetActivation)\n--- End of stack trace from previous location where exception was thrown ---\n   at Orleans.Internal.OrleansTaskExtentions.<ToTypedTask>g__ConvertAsync|4_0[T](Task`1 asyncTask)\n   at Orleans.Streams.PersistentStreamPullingAgent.DeliverBatchToConsumer(StreamConsumerData consumerData, IBatchContainer batch)\n   at Orleans.Internal.AsyncExecutorWithRetries.ExecuteWithRetriesHelper[T](Func`2 function, Int32 callCounter, Int32 maxNumSuccessTries, Int32 maxNumErrorTries, TimeSpan maxExecutionTime, DateTime startExecutionTime, Func`3 retryValueFilter, Func`3 retryExceptionFilter, IBackoffProvider onSuccessBackOff, IBackoffProvider onErrorBackOff)",
        "RemoteStackTraceString": null,
        "RemoteStackIndex": 0,
        "HResult": -2146233088,
        "HelpURL": null
      }
    ],
    "fields": {
      "Message": "Request S10.244.0.181:11111:332295207*stg/7872883655142754957/d6c34a8d@Sd6c34a8d->S10.244.0.181:11111:332295207*grn/E4F58DB0/9ccbb316@6285d9a0 #260",
      "Activation": "[Activation: S10.244.0.181:11111:332295207*grn/E4F58DB0/9ccbb316@6285d9a0 #GrainType=Grains.MemberGrain Placement=RandomPlacement State=Valid]",
      "EventId": {
        "Id": 101528,
        "Name": "Orleans.Messaging.Dispatcher.DetectedDeadlock"
      },
      "SourceContext": "Orleans.Messaging"
    }
  },
  "fields": {
    "@timestamp": [
      "2020-07-13T00:13:47.317Z"
    ]
  }
}

**Comment by sergeybykov** [comment]
What happens if you don't set `PerformDeadlockDetection` to `true`? I wonder if this is a false positive as deadlock detection is turned off by default.

**Comment by arbesoft** [comment]
If PerformDeadlockDetection is false then is working fine.

**Comment by Rohansi** [comment]
@sergeybykov I just ran into this as well. Are there any plans to correct this false positive? I'm using Orleans 3.4.1.

**Comment by sergeybykov** [comment]
@ReubenBond @benjaminpetit

**Comment by ReubenBond** [comment]
@Rohansi is this an issue specific to Azure Queues? It may be worth opening a new issue referencing this. It's difficult for us to track closed issues.

**Comment by Rohansi** [comment]
Thanks @ReubenBond, I created a new issue
